cpp/src/quantiles/quantiles_util.hpp

```diff
@@ -174,6 +174,71 @@ select_quantile(ValueAccessor get_value,
     }
 }
 
+template<typename Result, typename Iterator>
+CUDA_HOST_DEVICE_CALLABLE
+Result
+select_quantile_data(Iterator begin, size_type size, double q, interpolation interp)
+{
+    quantile_index idx(size, q);
+
+    switch (interp) {
+    case interpolation::LOWER:
+        return static_cast<Result>(*(begin + idx.lower));
```

Comment:
Can't the casting be done at the call site? That would be more obvious what's going on than an explicit template parameter.

Reply:
I don't believe so, due to underflow.

Reply:
Can you explain a little more? Looking briefly at the internals, I don't see any operation that needs to know `Result` as it's just internally casting to `Result` before returning. It seems like that casting could likewise just be done externally.

Reply:
https://github.com/rapidsai/cudf/pull/4292#discussion_r396598661

Reply:
Oh I think the part that is not obvious is that `intepolate::midpoint` has an overload which can handle `int64_t` underflow.